src/TorchSharp/JIT/ScriptModule.cs

```diff
@@ -163,9 +163,63 @@ public override void eval()
                 [DllImport("LibTorchSharp")]
                 static extern void THSJIT_Module_to_device(HType module, long deviceType, long deviceIndex);
 
+                [DllImport("LibTorchSharp")]
+                static extern void THSJIT_Module_to_device_dtype(HType module, sbyte dtype, long deviceType, long deviceIndex);
+
                 [DllImport("LibTorchSharp")]
                 static extern void THSJIT_Module_to_dtype(HType module, sbyte dtype);
 
+                public override nn.Module to(Device device, ScalarType dtype)
+                {
+                    if (device.type != DeviceType.CUDA) { device = new Device(device.type, -1); };
+
+                    if (device.type == DeviceType.CUDA && !torch.cuda.is_available()) throw new InvalidOperationException("CUDA is not available.");
+
+                    if (device.type != _deviceType || device.index != _deviceIndex) {
+
+                        InitializeDeviceType(device.type);
+                        THSJIT_Module_to_device_dtype(handle, (sbyte)dtype, (int)device.type, device.index);
+                        CheckForErrors();
+
+                        foreach (var (_, sm) in named_children()) sm.to(device.type, device.index);
+
+                        foreach (var field in GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance)) {
+
+                            var fieldName = field.Name;
+                            var value = field.GetValue(this);
+
+                            switch (value) {
+                            // This test must come before the Tensor test
+                            case Modules.Parameter param when dtype == param.dtype && device.type == param.device_type && device.index == param.device_index:
+                                continue;
+
+                            case Modules.Parameter param: {
+                                    var t = param.to(dtype, device);
+                                    t.retain_grad();
+                                    var p = new Modules.Parameter(t, param.requires_grad);
+                                    field.SetValue(this, p);
+                                    ConditionallyRegisterParameter(fieldName, p);
+                                    break;
+                                }
+
+                            case Tensor tensor when (device.type != tensor.device_type || device.index != tensor.device_index): {
+                                    var t = tensor.to(dtype, device);
+                                    field.SetValue(this, t);
+                                    ConditionallyRegisterBuffer(fieldName, t);
+                                    break;
+                                }
+                            }
+                        }
+
+                        _deviceType = device.type;
+                        _deviceIndex = device.index;
```

Comment:
Not related to this change, but is having device info cached in the object safe? Can I obtain two C# objects pointing to the same module? If so, and I move the module using one object, the other's state becomes inconsistent?

Reply:
That needs to be investigated. As you say, not related to this change, specifically.